### ISSUE TITLE
appimage-extract(-and-run) don't preserve file timestamps

### ISSUE THREAD

**Issue by realnc** [AppImage/AppImageKit#1151]
I wanted to fix this issue myself, seemed easy enough. Probably in `runtime.c` where this line is:

```c
chmod(prefixed_path_to_extract, st.st_mode);
```

Don't just apply permissions but also timestamps. However, since AppImageKit doesn't seem to be able to be compiled on modern systems, I can't add this myself.

As for why:

Not restoring timestamps breaks code that checks cache validity through timestamps. For example in one of my applications that I ship as an AppImage, I bundle LibVLC and it's plugins into the AppImage. Before the creation of the image, I generate the `usr/lib/vlc/plugins/plugins.dat` file. This speeds up VLC plugin loading. However, if the timestamps of the plugins changes, which happens when running the AppImage with `--appimage-extract-and-run`, LibVLC prints countless error messages about every plugin. One example:

> main libvlc error: stale plugins cache: modified /tmp/appimage_extracted_1e2e0dbe8053a3543648ca6c3c5a71fb/usr/lib/vlc/plugins/packetizer/libpacketizer_h264_plugin.so

As a result, the cache isn't used at all. This is not the end of the world, since it's a soft error, but still it would be nice if timestamps were preserved when extracting.

**Comment by probonopd** [comment]
Agree that extracting _should_ obey creation timestamps, sounds reasonable to me. What do others think?

**Comment by realnc** [comment]
According to `stat`, both `cp -a` as well as `tar` preserve modification time. They only change access time and "change" time (not sure what is.) Example:

```
$ stat some_file
File: some_file
  Size: 72965           Blocks: 144        IO Block: 4096   regular file
Device: 22h/34d Inode: 254422      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  realnc)   Gid: ( 1000/  realnc)
Access: 2021-09-24 00:09:33.648484504 +0300
Modify: 2019-04-06 03:41:26.000000000 +0300
Change: 2021-09-24 00:09:33.648484504 +0300
 Birth: -
```
"Access" and "Change" times were updated to current time. "Modify" time was preserved. Indeed, if I run the AppImage normally without extracting it, and while the application is open I either `cp -a /mnt/.mount_Application-ID` to somewhere else, or do `tar cf app.tar /mnt/.mount_Application-ID` and then untar it and run the application from the copy, it works correctly. Modification time is what matters.

**Comment by probonopd** [comment]
I think we should mirror the behavior of `tar`, do you agree?

**Comment by realnc** [comment]
> I think we should mirror the behavior of `tar`, do you agree?

Yes. I looked at some other formats as well, and they all do the same thing as tar.

**Comment by probonopd** [comment]
Great, so we know what we should do.
Pull requests would be highly welcome!

**Comment by realnc** [comment]
https://github.com/AppImage/AppImageKit/issues/1084#issuecomment-925952956
